{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "196b85b5-82cd-4e85-97c2-0be9de8af2c3",
   "metadata": {},
   "source": [
    "ASSIGNMENT : 12"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e57d152a-588a-4593-b5f2-d224b063e464",
   "metadata": {},
   "source": [
    "1.Explain why we have to use the Exception class while creating a Custom Exception.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b98791e-c626-457b-b66d-575c8384ed44",
   "metadata": {},
   "source": [
    "In Python programming language, it is recommended to subclass the built-in Exception class or one of its subclasses when creating a custom exception. This is because the Exception class is the base class for all built-in exceptions in Python, and provides a consistent interface and behavior for exceptions.\n",
    "\n",
    "When you subclass Exception or one of its subclasses, your custom exception inherits all the necessary methods and attributes of the base class, such as __str__() and args, allowing you to add your own specific functionality.\n",
    "\n",
    "In addition, Python's exception handling mechanism is designed to work with classes that inherit from Exception or its subclasses. When you raise an instance of a custom exception class, it can be caught and handled using the same syntax and mechanisms as built-in exceptions.\n",
    "\n",
    "Overall, by using the Exception class as the base class for custom exceptions in Python, you can ensure that your exceptions are well-behaved and can be integrated seamlessly into your codebase."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9abba8f-8758-407c-91f2-bfba71176900",
   "metadata": {},
   "outputs": [
    {
     "ename": "CustomException",
     "evalue": "Something went wrong.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCustomException\u001b[0m                           Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m         \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__init__\u001b[39m(message)\n\u001b[1;32m      4\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmessage \u001b[38;5;241m=\u001b[39m message\n\u001b[0;32m----> 5\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m CustomException(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSomething went wrong.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mCustomException\u001b[0m: Something went wrong."
     ]
    }
   ],
   "source": [
    "class CustomException(Exception):\n",
    "    def __init__(self, message):\n",
    "        super().__init__(message)\n",
    "        self.message = message\n",
    "raise CustomException(\"Something went wrong.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "789d6383-71dc-42bb-8180-85166ed333a8",
   "metadata": {},
   "source": [
    "2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ca07e4b9-4803-45f1-8d80-920959f40638",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ArithmeticError\n",
      "    FloatingPointError\n",
      "    OverflowError\n",
      "    ZeroDivisionError\n",
      "        DivisionByZero\n",
      "        DivisionUndefined\n",
      "    DecimalException\n",
      "        Clamped\n",
      "        Rounded\n",
      "            Underflow\n",
      "            Overflow\n",
      "        Inexact\n",
      "            Underflow\n",
      "            Overflow\n",
      "        Subnormal\n",
      "            Underflow\n",
      "        DivisionByZero\n",
      "        FloatOperation\n",
      "        InvalidOperation\n",
      "            ConversionSyntax\n",
      "            DivisionImpossible\n",
      "            DivisionUndefined\n",
      "            InvalidContext\n",
      "AssertionError\n",
      "AttributeError\n",
      "    FrozenInstanceError\n",
      "BlockingIOError\n",
      "BrokenPipeError\n",
      "BufferError\n",
      "ChildProcessError\n",
      "ConnectionAbortedError\n",
      "ConnectionError\n",
      "    BrokenPipeError\n",
      "    ConnectionAbortedError\n",
      "    ConnectionRefusedError\n",
      "    ConnectionResetError\n",
      "        RemoteDisconnected\n",
      "ConnectionRefusedError\n",
      "ConnectionResetError\n",
      "    RemoteDisconnected\n",
      "EOFError\n",
      "    IncompleteReadError\n",
      "OSError\n",
      "    ConnectionError\n",
      "        BrokenPipeError\n",
      "        ConnectionAbortedError\n",
      "        ConnectionRefusedError\n",
      "        ConnectionResetError\n",
      "            RemoteDisconnected\n",
      "    BlockingIOError\n",
      "    ChildProcessError\n",
      "    FileExistsError\n",
      "    FileNotFoundError\n",
      "    IsADirectoryError\n",
      "    NotADirectoryError\n",
      "    InterruptedError\n",
      "        InterruptedSystemCall\n",
      "    PermissionError\n",
      "    ProcessLookupError\n",
      "    TimeoutError\n",
      "    UnsupportedOperation\n",
      "    itimer_error\n",
      "    herror\n",
      "    gaierror\n",
      "    SSLError\n",
      "        SSLCertVerificationError\n",
      "        SSLZeroReturnError\n",
      "        SSLWantWriteError\n",
      "        SSLWantReadError\n",
      "        SSLSyscallError\n",
      "        SSLEOFError\n",
      "    Error\n",
      "        SameFileError\n",
      "    SpecialFileError\n",
      "    ExecError\n",
      "    ReadError\n",
      "    URLError\n",
      "        HTTPError\n",
      "        ContentTooShortError\n",
      "    BadGzipFile\n",
      "FileExistsError\n",
      "FileNotFoundError\n",
      "FloatingPointError\n",
      "OSError\n",
      "    ConnectionError\n",
      "        BrokenPipeError\n",
      "        ConnectionAbortedError\n",
      "        ConnectionRefusedError\n",
      "        ConnectionResetError\n",
      "            RemoteDisconnected\n",
      "    BlockingIOError\n",
      "    ChildProcessError\n",
      "    FileExistsError\n",
      "    FileNotFoundError\n",
      "    IsADirectoryError\n",
      "    NotADirectoryError\n",
      "    InterruptedError\n",
      "        InterruptedSystemCall\n",
      "    PermissionError\n",
      "    ProcessLookupError\n",
      "    TimeoutError\n",
      "    UnsupportedOperation\n",
      "    itimer_error\n",
      "    herror\n",
      "    gaierror\n",
      "    SSLError\n",
      "        SSLCertVerificationError\n",
      "        SSLZeroReturnError\n",
      "        SSLWantWriteError\n",
      "        SSLWantReadError\n",
      "        SSLSyscallError\n",
      "        SSLEOFError\n",
      "    Error\n",
      "        SameFileError\n",
      "    SpecialFileError\n",
      "    ExecError\n",
      "    ReadError\n",
      "    URLError\n",
      "        HTTPError\n",
      "        ContentTooShortError\n",
      "    BadGzipFile\n",
      "ImportError\n",
      "    ModuleNotFoundError\n",
      "    ZipImportError\n",
      "IndentationError\n",
      "    TabError\n",
      "IndexError\n",
      "InterruptedError\n",
      "    InterruptedSystemCall\n",
      "IsADirectoryError\n",
      "KeyError\n",
      "    NoSuchKernel\n",
      "    UnknownBackend\n",
      "LookupError\n",
      "    IndexError\n",
      "    KeyError\n",
      "        NoSuchKernel\n",
      "        UnknownBackend\n",
      "    CodecRegistryError\n",
      "MemoryError\n",
      "ModuleNotFoundError\n",
      "NameError\n",
      "    UnboundLocalError\n",
      "NotADirectoryError\n",
      "NotImplementedError\n",
      "    ZMQVersionError\n",
      "    StdinNotImplementedError\n",
      "OSError\n",
      "    ConnectionError\n",
      "        BrokenPipeError\n",
      "        ConnectionAbortedError\n",
      "        ConnectionRefusedError\n",
      "        ConnectionResetError\n",
      "            RemoteDisconnected\n",
      "    BlockingIOError\n",
      "    ChildProcessError\n",
      "    FileExistsError\n",
      "    FileNotFoundError\n",
      "    IsADirectoryError\n",
      "    NotADirectoryError\n",
      "    InterruptedError\n",
      "        InterruptedSystemCall\n",
      "    PermissionError\n",
      "    ProcessLookupError\n",
      "    TimeoutError\n",
      "    UnsupportedOperation\n",
      "    itimer_error\n",
      "    herror\n",
      "    gaierror\n",
      "    SSLError\n",
      "        SSLCertVerificationError\n",
      "        SSLZeroReturnError\n",
      "        SSLWantWriteError\n",
      "        SSLWantReadError\n",
      "        SSLSyscallError\n",
      "        SSLEOFError\n",
      "    Error\n",
      "        SameFileError\n",
      "    SpecialFileError\n",
      "    ExecError\n",
      "    ReadError\n",
      "    URLError\n",
      "        HTTPError\n",
      "        ContentTooShortError\n",
      "    BadGzipFile\n",
      "OverflowError\n",
      "PermissionError\n",
      "ProcessLookupError\n",
      "RecursionError\n",
      "ReferenceError\n",
      "RuntimeError\n",
      "    RecursionError\n",
      "    NotImplementedError\n",
      "        ZMQVersionError\n",
      "        StdinNotImplementedError\n",
      "    _DeadlockError\n",
      "    BrokenBarrierError\n",
      "    BrokenExecutor\n",
      "        BrokenThreadPool\n",
      "    SendfileNotAvailableError\n",
      "    ExtractionError\n",
      "    VariableError\n",
      "SyntaxError\n",
      "    IndentationError\n",
      "        TabError\n",
      "SystemError\n",
      "    CodecRegistryError\n",
      "TabError\n",
      "TimeoutError\n",
      "TypeError\n",
      "    FloatOperation\n",
      "    MultipartConversionError\n",
      "UnboundLocalError\n",
      "UnicodeDecodeError\n",
      "UnicodeEncodeError\n",
      "UnicodeError\n",
      "    UnicodeEncodeError\n",
      "    UnicodeDecodeError\n",
      "    UnicodeTranslateError\n",
      "UnicodeTranslateError\n",
      "ValueError\n",
      "    UnicodeError\n",
      "        UnicodeEncodeError\n",
      "        UnicodeDecodeError\n",
      "        UnicodeTranslateError\n",
      "    UnsupportedOperation\n",
      "    JSONDecodeError\n",
      "    SSLCertVerificationError\n",
      "    Error\n",
      "    UnsupportedDigestmodError\n",
      "    IllegalMonthError\n",
      "    IllegalWeekdayError\n",
      "    ParserError\n",
      "    ClassNotFound\n",
      "    ClipboardEmpty\n",
      "    MessageDefect\n",
      "        NoBoundaryInMultipartDefect\n",
      "        StartBoundaryNotFoundDefect\n",
      "        CloseBoundaryNotFoundDefect\n",
      "        FirstHeaderLineIsContinuationDefect\n",
      "        MisplacedEnvelopeHeaderDefect\n",
      "        MissingHeaderBodySeparatorDefect\n",
      "        MultipartInvariantViolationDefect\n",
      "        InvalidMultipartContentTransferEncodingDefect\n",
      "        UndecodableBytesDefect\n",
      "        InvalidBase64PaddingDefect\n",
      "        InvalidBase64CharactersDefect\n",
      "        InvalidBase64LengthDefect\n",
      "        HeaderDefect\n",
      "            InvalidHeaderDefect\n",
      "            HeaderMissingRequiredValue\n",
      "            NonPrintableDefect\n",
      "            ObsoleteHeaderDefect\n",
      "            NonASCIILocalPartDefect\n",
      "            InvalidDateDefect\n",
      "    MacroToEdit\n",
      "    InvalidFileException\n",
      "    UnequalIterablesError\n",
      "    InvalidVersion\n",
      "    _InvalidELFFileHeader\n",
      "    InvalidWheelFilename\n",
      "    InvalidSdistFilename\n",
      "    InvalidSpecifier\n",
      "    InvalidMarker\n",
      "    UndefinedComparison\n",
      "    UndefinedEnvironmentName\n",
      "    InvalidRequirement\n",
      "        RequirementParseError\n",
      "    InvalidVersion\n",
      "ZeroDivisionError\n",
      "    DivisionByZero\n",
      "    DivisionUndefined\n"
     ]
    }
   ],
   "source": [
    "exceptions = [eval(exp) for exp in dir(__builtins__) if exp.endswith('Error')]\n",
    "\n",
    "def print_hierarchy(cls, indent=0):\n",
    "    print(' ' * indent + cls.__name__)\n",
    "    for subclass in cls.__subclasses__():\n",
    "        print_hierarchy(subclass, indent + 4)\n",
    "\n",
    "for exception in exceptions:\n",
    "    print_hierarchy(exception)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98c6401b-35ae-4bad-8277-7c1649d9d2c2",
   "metadata": {},
   "source": [
    "3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "523a6b28-5b08-42ab-9035-9046d56d42d7",
   "metadata": {},
   "source": [
    "In Python, the ArithmeticError class is the base class for all errors that occur during arithmetic operations. It has several subclasses that represent specific types of arithmetic errors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b16a9a6-9a0f-4d60-89ad-2b19d87bae91",
   "metadata": {},
   "source": [
    "ZeroDivisionError: This error is raised when you try to divide a number by zero. For example:\n",
    "x = 10,\n",
    "y = 0,\n",
    "z = x / y  # Raises ZeroDivisionError\n",
    "\n",
    "OverflowError: This error is raised when a calculation produces a result that is too large to be represented as a number in Python. For example:\n",
    "x = 2 ** 1000  # Raises OverflowError\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ead9a30a-418b-4c12-971e-8496a6de7346",
   "metadata": {},
   "source": [
    "4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d845c2b-f1ac-47eb-8e7d-84779d94fbee",
   "metadata": {},
   "source": [
    "In Python, the LookupError class is the base class for all errors that occur when trying to access a non-existent item in a sequence or mapping. It has several subclasses that represent specific types of lookup errors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f58fa7df-27ee-4164-a6e7-87cf00b5177e",
   "metadata": {},
   "source": [
    "KeyError: This error is raised when you try to access a non-existent key in a dictionary. \n",
    "my_dict = {\"a\": 1, \"b\": 2, \"c\": 3},\n",
    "value = my_dict[\"d\"]  # Raises KeyError\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38f0ed31-1d26-4832-bb64-fc4abc6f1eaf",
   "metadata": {},
   "source": [
    "IndexError: This error is raised when you try to access a non-existent index in a list or other sequence. For example:\n",
    "    my_list = [1, 2, 3],\n",
    "value = my_list[3]  # Raises IndexError\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3a6c01e-01a5-4e74-bda2-27bda7d607f5",
   "metadata": {},
   "source": [
    "5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ef67197-afe9-4fff-9c39-caafe180c167",
   "metadata": {},
   "source": [
    "In Python, ImportError is an exception that is raised when you try to import a module or package that cannot be found or loaded. This can happen for a variety of reasons, such as a misspelled module name, a missing or corrupted module file, or a problem with the Python path."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4449e785-2fe8-461c-8a0c-03830cc58c32",
   "metadata": {},
   "source": [
    "In Python 3.6 and later versions, ModuleNotFoundError is a subclass of ImportError that is raised when a module or package cannot be found during an import. It has the same behavior as ImportError, but provides a more specific error message to help diagnose the problem."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e34c27a-80af-4840-8c0f-2cb373d207d7",
   "metadata": {},
   "source": [
    "6. List down some best practices for exception handling in python. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a98a988d-df3b-495e-873f-31b5490018c2",
   "metadata": {},
   "source": [
    "Be specific: When catching exceptions, it's best to be as specific as possible. Rather than catching a broad exception like Exception, try to catch more specific exceptions like ValueError or TypeError. This can help you identify and diagnose problems more easily, and can also help prevent unintended consequences.\n",
    "\n",
    "Use try/except blocks sparingly: While try/except blocks can be useful for handling exceptions, they can also make code harder to read and understand. In general, it's best to only use try/except blocks when you're dealing with code that is likely to raise an exception, or when you need to handle exceptions in a specific way.\n",
    "\n",
    "Use finally blocks for cleanup: If you need to perform cleanup tasks (such as closing a file or database connection) regardless of whether an exception was raised, you can use a finally block. Code inside the finally block will always be executed, regardless of whether an exception was raised or not.\n",
    "\n",
    "Don't catch exceptions you can't handle: If you catch an exception but can't actually handle it, you're just hiding the problem from yourself or others. It's generally better to let the exception propagate up the call stack, where it can be handled more appropriately.\n",
    "\n",
    "Use context managers: Context managers (such as the with statement) are a good way to ensure that resources are properly cleaned up, even if an exception is raised. For example, you can use a context manager to open a file, and it will automatically be closed when you're done, even if an exception is raised.\n",
    "\n",
    "Document your exceptions: When defining your own exceptions, be sure to document what they represent and when they are raised. This can help other developers understand your code and use your exceptions correctly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e735b2ed-859b-4af6-b7c4-4455b054a602",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
